src/charm.py

```diff
@@ -1,24 +1,41 @@
 #!/usr/bin/env python3
-# Copyright 2021 Canonical Ltd.
+# Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
 import logging
 from random import choices
 from string import ascii_uppercase, digits
 
-from oci_image import OCIImageResource, OCIImageResourceError
+from charmed_kubeflow_chisme.exceptions import ErrorWithStatus
+from charmed_kubeflow_chisme.pebble import update_layer
+from charms.observability_libs.v1.kubernetes_service_patch import KubernetesServicePatch
+from lightkube.models.core_v1 import ServicePort
 from ops.charm import CharmBase
 from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import Layer
 from serialized_data_interface import NoCompatibleVersions, NoVersionsListed, get_interfaces
 
 
 class Operator(CharmBase):
     def __init__(self, *args):
         super().__init__(*args)
 
-        self.log = logging.getLogger(__name__)
-        self.image = OCIImageResource(self, "oci-image")
+        self.logger = logging.getLogger(__name__)
+        self._container_name = "oidc-authservice"
+        self._container = self.unit.get_container(self._container_name)
+        self.pebble_service_name = "oidc-authservice"
+        self._http_port = self.model.config["port"]
+
+        http_port = ServicePort(int(self._http_port), name="http-port")
+        self.service_patcher = KubernetesServicePatch(
+            self,
+            [http_port],
+        )
+
+        self.public_url = self.model.config["public-url"]
+        if not self.public_url.startswith(("http://", "https://")):
+            self.public_url = f"http://{self.public_url}"
 
         for event in [
             self.on.start,
```

Comment:
Can we handle upgrade with this setup or do we need separate upgrade event handler?